silkworm/db/snapshots/bittorrent/web_seed_client.cpp

```
@@ -0,0 +1,201 @@
+
+//! HTTP target for Manifest file containing the list of available snapshot files
+static constexpr std::string_view kManifestTarget{"/manifest.txt"sv};
+
+//! The BitTorrent file extension for torrent files
+static constexpr std::string_view kTorrentExtension{".torrent"sv};
+
+//! The size of BitTorrent file extension
+static const std::size_t kTorrentExtensionSize{std::strlen(kTorrentExtension.data())};
+
+//! Timeout for parallel async download of manifest files in msec
+static const std::chrono::milliseconds kParallelManifestDownloadTimeout{60'000};
+
+//! Timeout for parallel async download of torrent files in msec
+static const std::chrono::milliseconds kParallelTorrentDownloadTimeout{120'000};
+
+WebSeedClient::WebSeedClient(std::vector<std::string>&& url_seeds, const PreverifiedList& preverified)
+    : WebSeedClient(std::make_unique<WebSession>(), std::move(url_seeds), preverified) {}
+
+WebSeedClient::WebSeedClient(std::unique_ptr<WebSession>&& web_session,
+                             std::vector<std::string>&& url_seeds,
+                             const PreverifiedList& preverified)
+    : url_seeds_{std::move(url_seeds)},
+      preverified_{preverified},
+      web_session_{std::move(web_session)} {}
+
+Task<TorrentInfoPtrList> WebSeedClient::discover_torrents(bool fail_fast) {
+    torrents_by_provider_.clear();
+    co_await build_list_of_torrents(fail_fast);
+    co_return co_await download_and_filter_all_torrents();
+}
+
+Task<void> WebSeedClient::build_list_of_torrents(bool fail_fast) {
+    using namespace concurrency::awaitable_wait_for_one;
+
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-capturing-lambda-coroutines)
+    auto build_list_of_torrents_factory = [this, fail_fast](size_t index) -> Task<void> {
+        try {
+            const auto& provider_url = url_seeds_[index];
+            co_await build_list_of_torrents(provider_url);
+        } catch (const std::exception&) {
+            if (fail_fast) throw;
```

if fail_fast = false, we shouldn't ignore the exception